src/main/java/org/opensearch/neuralsearch/search/collector/HybridCollapsingTopDocsCollector.java

```diff
@@ -0,0 +1,316 @@
+                        LeafFieldComparator[] leafFieldComparators = compoundScores[i].getComparators(context);
+                        int[] reverseMuls = compoundScores[i].getReverseMul();
+                        if (leafFieldComparators.length == 1) {
+                            reverseMulMap.put(groupSelector.copyValue(), reverseMuls[0]);
+                            comparators[i] = leafFieldComparators[0];
+                        } else {
+                            reverseMulMap.put(groupSelector.copyValue(), 1);
+                            comparators[i] = new MultiLeafFieldComparator(leafFieldComparators, reverseMuls);
+                        }
+                        comparators[i].setScorer(compoundQueryScorer);
+                    }
+                    comparatorsMap.put(groupSelector.copyValue(), comparators);
+                    initializeLeafComparatorsPerSegmentOnceMap.put(groupSelector.copyValue(), false);
+                }
+                totalHitCount++;
+                hitsThresholdChecker.incrementHitCount();
+
+                if (hitsThresholdChecker.isThresholdReached()) {
+                    setTotalHitsRelation(TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO);
+                    log.info("Terminating collection early as specified hits threshold is reached");
+                    throw new CollectionTerminatedException();
+                }
+
+                for (int i = 0; i < subScoresByQuery.length; i++) {
+                    float score = subScoresByQuery[i];
+                    if (score == 0) {
+                        continue;
+                    }
+
+                    boolean[] queueFullArray = queueFullMap.get(groupValue);
+                    if (queueFullArray[i]) {
+                        LeafFieldComparator[] comparators = comparatorsMap.get(groupValue);
+                        if (reverseMulMap.get(groupValue) * comparators[i].compareBottom(doc) > 0) {
+                            FieldValueHitQueue.Entry[] fieldValueLeafTrackers = fieldValueLeafTrackersMap.get(groupValue);
+                            FieldValueHitQueue<FieldValueHitQueue.Entry>[] compoundScores = groupQueueMap.get(groupValue);
+                            comparators[i].copy(fieldValueLeafTrackers[i].slot, doc);
+
+                            fieldValueLeafTrackers[i].doc = docBase + doc;
+                            fieldValueLeafTrackers[i] = compoundScores[i].updateTop();
+                            comparators[i].setBottom(fieldValueLeafTrackers[i].slot);
+
+                            comparatorsMap.put(groupSelector.copyValue(), comparators);
+                            fieldValueLeafTrackersMap.put(groupSelector.copyValue(), fieldValueLeafTrackers);
+                            groupQueueMap.put(groupSelector.copyValue(), compoundScores);
+                        }
+                    } else {
+                        int[] collectedHitsForCurrentSubQuery = collectedHitsPerSubQueryMap.get(groupValue);
```

Comment:
There is lot happening below. I am not able to understand it completely. Can you take reference from HybridTopDocFieldCollector to see how we can create updateBottom like methods to segregate it properly.